File: examples/raw-python/payxcommerce.py
"""Dependency-free PayXCommerce API helpers for raw Python integrations."""

from __future__ import annotations

import hashlib
import hmac
import json
import re
import secrets
import time
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Optional

DEFAULT_BASE_URL = "https://payxcommerce.com/api/v1"
TOKEN_SCOPE = "payment_requests.write transactions.read balances.read refunds.write"


def default_ipn_events() -> list[str]:
    return [
        "payment.succeeded",
        "payment.failed",
        "payment.cancelled",
        "payment.expired",
        "refund.succeeded",
        "payment.refunded",
        "chargeback.created",
        "dispute.created",
    ]


def json_body(payload: Optional[dict[str, Any]]) -> str:
    return "" if payload is None else json.dumps(payload, separators=(",", ":"), ensure_ascii=False)


def hmac_headers(public_key: str, secret_key: str, body: str, idempotency_key: Optional[str] = None) -> dict[str, str]:
    timestamp = str(int(time.time()))
    nonce = secrets.token_hex(16)
    signature = hmac.new(secret_key.encode(), f"{timestamp}.{nonce}.{body}".encode(), hashlib.sha256).hexdigest()
    headers = {
        "X-PXC-Public-Key": public_key,
        "X-PXC-Timestamp": timestamp,
        "X-PXC-Nonce": nonce,
        "X-PXC-Signature": signature,
    }
    if idempotency_key:
        headers["Idempotency-Key"] = idempotency_key
    return headers


def json_request(method: str, url: str, headers: Optional[dict[str, str]] = None, payload: Optional[dict[str, Any]] = None) -> dict[str, Any]:
    body = json_body(payload)
    request_headers = {"Accept": "application/json", "Content-Type": "application/json"}
    request_headers.update(headers or {})
    request = urllib.request.Request(url, data=body.encode() if body else None, headers=request_headers, method=method.upper())

    try:
        with urllib.request.urlopen(request, timeout=30) as response:
            response_body = response.read().decode()
            return {"status": response.status, "body": parse_json(response_body)}
    except urllib.error.HTTPError as error:
        response_body = error.read().decode()
        return {"status": error.code, "body": parse_json(response_body)}


def parse_json(value: str) -> Any:
    if value == "":
        return {}
    try:
        return json.loads(value)
    except json.JSONDecodeError:
        return value


def oauth_client_credentials(base_url: str, client_id: str, client_secret: str, scope: str = TOKEN_SCOPE) -> dict[str, Any]:
    return json_request(
        "POST",
        endpoint(base_url, "/oauth/token"),
        payload={"grant_type": "client_credentials", "client_id": client_id, "client_secret": client_secret, "scope": scope},
    )


def verify_webhook(raw_body: str, headers: dict[str, str], webhook_secret: str, tolerance_seconds: int = 300) -> dict[str, Any]:
    event_id = header(headers, "X-PXC-Event-ID")
    timestamp = header(headers, "X-PXC-Timestamp")
    signature = header(headers, "X-PXC-Signature")

    if not event_id or not timestamp or not signature:
        raise ValueError("Missing PayXCommerce webhook signature headers.")
    if not timestamp.isdigit() or abs(int(time.time()) - int(timestamp)) > tolerance_seconds:
        raise ValueError("Invalid or expired PayXCommerce webhook timestamp.")

    payload = json.loads(raw_body)
    canonical_body = json.dumps(payload, separators=(",", ":"), ensure_ascii=False)
    expected = hmac.new(webhook_secret.encode(), f"{event_id}.{canonical_body}".encode(), hashlib.sha256).hexdigest()
    if not hmac.compare_digest(expected, signature):
        raise ValueError("Invalid PayXCommerce webhook signature.")
    return payload


def header(headers: dict[str, str], name: str) -> str:
    normalized = name.lower()
    for key, value in headers.items():
        if key.lower() == normalized:
            return value
    return ""


def endpoint(base_url: str, path: str) -> str:
    return base_url.rstrip("/") + "/" + path.lstrip("/")


def print_response(response: dict[str, Any]) -> None:
    print(f"HTTP Status: {response['status']}")
    print(json.dumps(response["body"], indent=2, ensure_ascii=False))


def redact(message: str) -> str:
    return re.sub(
        r"(secret|token|signature|authorization|password|key|client_secret|secret_key|webhook_secret)([^\s:=]*)?([:=]\s*)?([A-Za-z0-9_\-.!@$%^&*+/=]+)",
        r"\1\2\3[redacted]",
        message,
        flags=re.IGNORECASE,
    )

File: examples/raw-nodejs/payxcommerce.js
'use strict';

const crypto = require('crypto');
const http = require('http');
const https = require('https');

const DEFAULT_BASE_URL = 'https://payxcommerce.com/api/v1';
const TOKEN_SCOPE = 'payment_requests.write transactions.read balances.read refunds.write';

function defaultIpnEvents() {
  return [
    'payment.succeeded',
    'payment.failed',
    'payment.cancelled',
    'payment.expired',
    'refund.succeeded',
    'payment.refunded',
    'chargeback.created',
    'dispute.created'
  ];
}

function jsonBody(payload) {
  return payload == null ? '' : JSON.stringify(payload);
}

function hmacHeaders(publicKey, secretKey, body, idempotencyKey) {
  const timestamp = Math.floor(Date.now() / 1000).toString();
  const nonce = crypto.randomBytes(16).toString('hex');
  const signature = crypto.createHmac('sha256', secretKey).update(`${timestamp}.${nonce}.${body}`).digest('hex');
  const headers = {
    'X-PXC-Public-Key': publicKey,
    'X-PXC-Timestamp': timestamp,
    'X-PXC-Nonce': nonce,
    'X-PXC-Signature': signature
  };
  if (idempotencyKey) {
    headers['Idempotency-Key'] = idempotencyKey;
  }
  return headers;
}

function jsonRequest(method, url, headers = {}, payload = null) {
  const body = jsonBody(payload);
  const target = new URL(url);
  const transport = target.protocol === 'http:' ? http : https;
  const requestHeaders = Object.assign({ Accept: 'application/json', 'Content-Type': 'application/json' }, headers);
  if (body) {
    requestHeaders['Content-Length'] = Buffer.byteLength(body);
  }

  return new Promise((resolve, reject) => {
    const request = transport.request({
      method: method.toUpperCase(),
      hostname: target.hostname,
      port: target.port || undefined,
      path: target.pathname + target.search,
      headers: requestHeaders,
      timeout: 30000
    }, (response) => {
      let responseBody = '';
      response.setEncoding('utf8');
      response.on('data', (chunk) => { responseBody += chunk; });
      response.on('end', () => resolve({ status: response.statusCode, body: parseJson(responseBody) }));
    });

    request.on('timeout', () => request.destroy(new Error('PayXCommerce request timed out.')));
    request.on('error', reject);
    if (body) {
      request.write(body);
    }
    request.end();
  });
}

function parseJson(value) {
  if (!value) {
    return {};
  }
  try {
    return JSON.parse(value);
  } catch (error) {
    return value;
  }
}

function endpoint(baseUrl, path) {
  return `${baseUrl.replace(/\/+$/, '')}/${path.replace(/^\/+/, '')}`;
}

async function oauthClientCredentials(baseUrl, clientId, clientSecret, scope = TOKEN_SCOPE) {
  return jsonRequest('POST', endpoint(baseUrl, '/oauth/token'), {}, {
    grant_type: 'client_credentials',
    client_id: clientId,
    client_secret: clientSecret,
    scope
  });
}

function verifyWebhook(rawBody, headers, webhookSecret, toleranceSeconds = 300) {
  const eventId = header(headers, 'X-PXC-Event-ID');
  const timestamp = header(headers, 'X-PXC-Timestamp');
  const signature = header(headers, 'X-PXC-Signature');
  if (!eventId || !timestamp || !signature) {
    throw new Error('Missing PayXCommerce webhook signature headers.');
  }
  if (!/^\d+$/.test(timestamp) || Math.abs(Math.floor(Date.now() / 1000) - Number(timestamp)) > toleranceSeconds) {
    throw new Error('Invalid or expired PayXCommerce webhook timestamp.');
  }

  const payload = JSON.parse(rawBody);
  const canonicalBody = JSON.stringify(payload);
  const expected = crypto.createHmac('sha256', webhookSecret).update(`${eventId}.${canonicalBody}`).digest('hex');
  if (expected.length !== signature.length) {
    throw new Error('Invalid PayXCommerce webhook signature.');
  }
  if (!crypto.timingSafeEqual(Buffer.from(expected), Buffer.from(signature))) {
    throw new Error('Invalid PayXCommerce webhook signature.');
  }
  return payload;
}

function header(headers, name) {
  const normalized = name.toLowerCase();
  for (const [key, value] of Object.entries(headers)) {
    if (key.toLowerCase() === normalized) {
      return String(value || '');
    }
  }
  return '';
}

function printResponse(response) {
  console.log(`HTTP Status: ${response.status}`);
  console.log(JSON.stringify(response.body, null, 2));
}

function redact(message) {
  return String(message).replace(/(secret|token|signature|authorization|password|key|client_secret|secret_key|webhook_secret)([^\s:=]*)?([:=]\s*)?([A-Za-z0-9_\-.!@$%^&*+/=]+)/gi, '$1$2$3[redacted]');
}

module.exports = {
  DEFAULT_BASE_URL,
  TOKEN_SCOPE,
  defaultIpnEvents,
  endpoint,
  hmacHeaders,
  jsonBody,
  jsonRequest,
  oauthClientCredentials,
  printResponse,
  redact,
  verifyWebhook
};
